### ISSUE TITLE
Name error: type iterator/2 undefined

### ISSUE THREAD

**Issue by skogsbaer** [etylizer/etylizer#154]
Test file:

```erlang
-module(test).

-compile(export_all).
-compile(nowarn_export_all).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, maps:to_list(Map)).
```

**Comment by skogsbaer** [comment]
Just verified: the bug is still in main

**Comment by skogsbaer** [comment]
The problem is in `ast_to_erlang_ty`, when transforming a named type. The named type is `iterator/2`, the reference is contained in `maps.erl`. The symtab contains an entry for `iterator/2`, but qualified as `maps:iterator/2`.

We need to fix this bug, it will hit us often when doing the case study.

**Comment by skogsbaer** [comment]
The symtab contains unqualified entries for the current module but qualified entries for all other modules. One solution could be:

- Use qualified names only as keys in the symtab.
- Transform unqualified names in source programs to qualified names in AST transform.

@albsch sounds good?

**Comment by skogsbaer** [comment]
A somewhat minimal testcase:

File `m1.erl`

```erlang
-module(m1).

-compile(export_all).
-compile(nowarn_export_all).

-type blub() :: integer().

-spec foo(blub()) -> integer().
foo(X) -> X.
```

File `m2.erl`

```erlang
-module(m2).

-compile(export_all).
-compile(nowarn_export_all).

-spec bar() -> integer().
bar() -> m1:foo(1).
```

Then compile `m2.erl`. It fails with `./m1.erl:8:11: Name error: type blub/0 undefined`

**Comment by skogsbaer** [comment]
Here is an example that demonstrates that qualified all symbtab entries with module names is not enough. Instead, we have to use file paths.

Consider there are two directories A and B with the following files:

```
A/m1.erl A/m2.erl
B/m2.erl
```

We type check like this: ety -S A B/m2.erl

Content of `B/m2.erl`

```erlang
-module(m2).

-compile(export_all).
-compile(nowarn_export_all).

% Type m2:blub() in directory B is never used.
-type blub() :: string().

-spec bar() -> integer().
bar() -> m1:foo(1).
```

Content of `A/m2.erl`:

```erlang
-module(m2).

-compile(export_all).
-compile(nowarn_export_all).

% This is m2:blub in A/m2.erl
-type blub() :: integer().
```

So there are two definitions for `m2:blub()`: one in `A/m2.erl`, the other in `B/m2.erl`

Content of `A/m1.erl`

```erlang
-module(m1).

-compile(export_all).
-compile(nowarn_export_all).

% m2:blub() refers to A/m2.erl (not B/m2.erl)
-spec foo(m2:blub()) -> integer().
foo(X) -> X.
```

From my intution, reference `m2:blub()` in `A/m1.erl` should refer to `blub` in `A/m2.erl` because this file is "closer" than `B/m2.erl`. Hence, the program should type check. It currently fails with

```
A/m1.erl:8:11: Type error: expression failed to type check
%    8| foo(X) -> X.
```

Dialyzer does not allow such duplicated modules in the same project. But dialyzer allows to use the same module name as same library, effectively overwriting the module from the library.

I could not find documentation that explains how erlang modules should be resolved and whether it is possible to have two modules with the same name in one project. 

@albsch @bieniusa any thoughts on this?

**Comment by bieniusa** [comment]
The unwritten(?) rule I learned is that you should not have modules with the same name in one project. Though this should eventually be fixed in the language imho, I wouldn't try to solve it now here. 
This seems to describe the intended behavior:[https://erlang.org/workshop/2002/Carlsson.pdf](https://erlang.org/workshop/2002/Carlsson.pdf)
Apparently, clashes need to be identified manually via `code:clash()`

**Comment by skogsbaer** [comment]
@bieniusa does this mean that user projects cannot use module names that are already used in the standard library?

**Comment by bieniusa** [comment]
One can change the order of loads and/or exclude paths. However, module names from the standard library should generally be avoided.

**Comment by albsch** [comment]
Folders in an Erlang project is also a somewhat recent development, usually every Erlang source file lives in the same folder (or according to the [code loading strategy](https://www.erlang.org/doc/system/system_principles.html#code-loading-strategy), in versioned folders). While you can load multiple versions of Erlang modules at the same time for hot-code upgrades, this feature is not intended to have multiple Erlang modules loaded which do different things. So no, user projects cannot use module names that are already used in the standard library. Further, you cannot use module names that appear in your dependencies (e.g. try creating an empty `eunit.erl` file in the project).

But you're right, I couldn't find anything related to this in the official documentation.

**Comment by skogsbaer** [comment]
Fixed in #170
